non-normative/recipes.adoc

```diff
@@ -0,0 +1,32 @@
+[[recipe-brs-i-guidance]]
+===== BRS-I Recipe Guidance
+
+Systems compliant to BRS-I can successfully boot an existing generic
+operating system image without system-specific customizations, yet
+this might result in an unoptimized experience and non-functioning
+I/O devices until further software updates are activated.
+
+The best analogy would be a typical Intel Architecture motherboard from
+the early 2000s: you could install an OS on it, but the built-in graphics
+might be low-resolution and the sound, built-in network port or power
+management might not work out of the box. You could subsequently load
+the right drivers from the media coming with the board or fetch newest
+ones using a well-supported network adapter.
+
+Heterogeneous performance harts (e.g. mix of "performance" and "efficiency"
+harts) is a great example of a feature outside the current scope of BRS-I,
+yet with potential for adverse effects to a generic operating system
+written against the BRS-I spec. Such a feature does not have a standard
+way of being detected by an OS, leading to scheduling anomalies in
+operating systems that are not specially adapted. Consider two identical
+threads that, unbeknownst to the OS, get scheduled on harts with different
+performance characteristics. This might see dramatic difference in forward
+progress being made, with unexpected delays, timeouts or even crashes
+posible. Worse, there could be little to no system messages pointing to
+the cause of such behavior. Thus, such a feature would need to be made
+*opt-in* for BRS-I compliance. This could be done implicitly via loading
+of system-specific drivers, or explicitly via firmware setup and
+configuration utilities. The actual meaning of the default (compatible to
+BRS-I) configuration would be highly specific to the vendor. In this example,
+it could capping all harts at the same performance level, or it could mean
+disabling efficiency harts.
```

Comment:
Does that mean they disable some harts for compliance but enable them in the versions they ship it to the customer ?

Reply:
It's an example. All I am saying is that if you have behavior that impacts a generic OS image in the manner defined in this commit, requiring additional software to make things work in a non-anomalous fashion, then such behavior must be only enabled (in whatever fashion) after the supporting software is installed.

For this example, you could:
-  have a UEFI setting that allows disabling E cores (that could be disabled once the relevant non-standard system software is installed)
- cap performance level for all harts

...for any approach taken, the feature could be unlocked via a hardware access, SBI call, UEFI NVRAM variable access, etc from the system software itself (or if you can't - manually in UEFI setup).

Again the point is that if you have a feature that prevents a generic BRS-I compliant OS from booting correctly (within the boundaries identified), then yes your feature is only compliant to BRS-I if it is opt-in instead of forced on every boot.

Reply:
Ahh got it. You mean the feature must be enabled at runtime rather than at boot time. In that way, compliance is not affected.

Reply:
Yes, I think we're on the same page now.

(to be pedantic, the feature *could* be enabled at boot time - it's a choice, for example by way of an OS driver setting a UEFI variable so that subsequent boots see the feature enabled, or an operator being asked to enable something in the firmware setup dialogs... the key aspect is that it's *not* a default configuration... going into UEFI setup and "pick defaults" should place the machine into a state where BRS-I compliant OSes boot)